plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+    coordinateHistoricExport: (update?: CoordinationUpdate) => Promise<void>
+    calculateCoordination: (
+        params: ExportParams,
+        done: Array<ISOTimestamp>,
+        running: Array<ISOTimestamp>
+    ) => Promise<CoordinationUpdate>
+    getExportDateRange: (params: ExportParams) => Array<[ISOTimestamp, ISOTimestamp]>
+    progressBar: (progress: number, length?: number) => string
+}
+
+export type ExportHistoricalEventsUpgradeV2 = Plugin<{
+    global: {
+        _testFunctions: TestFunctions
+    }
+}>
+
+export interface ExportHistoricalEventsJobPayload {
+    // Current cursor to what's being exported
+    timestampCursor: number
+
+    // The lower and upper bound of the timestamp interval to be processed
+    startTime: number
+    endTime: number
+
+    // The offset *within* a given timestamp interval
+    offset: number
+
+    // how many retries a payload has had (max = 15)
+    retriesPerformedSoFar: number
+
+    // used for ensuring only one "export task" is running if the server restarts
+    exportId: number
+
+    // Time frame to fetch events for.
+    fetchTimeInterval: number
+
+    // Key to report export status to
+    statusKey: string
+}
+
+export interface ExportHistoricalEventsUIPayload {
+    // Only set starting export from UI
+    parallelism?: number
+    dateFrom?: string
+    dateTo?: string
+}
+
+export interface ExportParams {
+    id: number
+    parallelism: number
+    dateFrom: ISOTimestamp
+    dateTo: ISOTimestamp
+}
+
+interface CoordinationPayload {
+    running?: Array<ISOTimestamp>
+    done?: Array<ISOTimestamp>
+    progress?: number
+}
+
+interface CoordinationUpdate {
+    hasChanges: boolean
+    done: Array<ISOTimestamp>
+    running: Array<ISOTimestamp>
+    toStartRunning: Array<[ISOTimestamp, ISOTimestamp]>
+    progress: number
+    exportIsDone: boolean
+}
+
+interface ExportDateStatus {
+    done: boolean
+    progress: number
+}
+
+export function addHistoricalEventsExportCapabilityV2(
+    hub: Hub,
+    pluginConfig: PluginConfig,
+    response: PluginConfigVMInternalResponse<PluginMeta<ExportHistoricalEventsUpgradeV2>>
+): void {
+    const { methods, tasks, meta } = response
+
+    const currentPublicJobs = pluginConfig.plugin?.public_jobs || {}
+
+    if (!(INTERFACE_JOB_NAME in currentPublicJobs)) {
+        hub.promiseManager.trackPromise(hub.db.addOrUpdatePublicJob(pluginConfig.plugin_id, INTERFACE_JOB_NAME, {}))
+    }
+
+    const oldRunEveryMinute = tasks.schedule.runEveryMinute?.exec
+
+    tasks.job[INTERFACE_JOB_NAME] = {
+        name: INTERFACE_JOB_NAME,
+        type: PluginTaskType.Job,
+        // TODO: Accept timestamp as payload
+        exec: async (payload: ExportHistoricalEventsUIPayload) => {
+            // only let one export run at a time
+            const alreadyRunningExport = await getExportParameters()
+            if (!!alreadyRunningExport) {
```

Comment:
probably acceptable for our current use case, but this is susceptible to a race condition

Reply:
I think we can ignore this one and assume UI jobs are triggered by users -> low volume.